### PR TITLE
Use window.Misk.History in Tab React Routers

### PR DESCRIPTION
**Pull request by adrw**
Tab sub-routers were not re-rendering on location change. Now, all React Routers in all tabs are initialized from the same History object that will live in `window.Misk.History`.

Other Changes
- Dashboard logo/title in NavTopbar now is a `<Link>`
- Config tab now properly hides when not on a matching route
- Config tab now reloads on soft refreshes
- Config tab separated out from data loading logic into TabContainer for data control and ConfigComponent for view
- Javascript Multibinder allows for any tab to use a common `multibind(binder, { ... bound object})` function to add objects to a multibinder in `window.Misk.Binder`

### CODE REVIEW

**Comment by terrilldent** [comment]
It would be really nice to see smaller PRs.

File: misk/src/main/kotlin/misk/web/proxy/WebProxyEntry.kt [hunk 1]
```diff
@@ -38,5 +38,6 @@ fun WebProxyEntry(
   url_path_prefix: String,
   web_proxy_url: String
 ) : WebProxyEntry {
+//  TODO(adrw) update all HTTPUrl.parse -> get (no bang bang required) https://github.com/square/misk/issues/419
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/web/proxy/WebProxyEntry.kt]: I’ll show you tools so it’s easier to just make fixes like this vs. writing comments

TODOs should be hard!

File: misk/web/tabs/config/src/components/ConfigComponent.tsx [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+import * as React from "react"
+import styled from "styled-components"
+import { IConfigFile } from "../containers/TabContainer"
+
+interface IConfigProps {
+  files: IConfigFile[]
+  status: string
+}
+
+const Container = styled.div`
```
Comment (swankjesse) [inline misk/web/tabs/config/src/components/ConfigComponent.tsx]: this inline CSS syntax is neat

File: misk/web/tabs/config/src/components/ConfigComponent.tsx [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+  margin-left: 180px;
+  margin-top: 20px;
+`
+
+export default class ConfigComponent extends React.PureComponent<IConfigProps> {
+  indent(spaces: number) {
+    let final = ""
```
Comment (swankjesse) [inline misk/web/tabs/config/src/components/ConfigComponent.tsx]: is `final` a convention? We use `result` in Kotlin  . . . (`final` is surprising ’cause it gets replaced)

File: misk/web/tabs/config/src/components/ConfigComponent.tsx [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+    for (let i = 0; i < spaces; i++) {
+      final += "\ "
+    }
+    return final
+  }
+
+  toYaml(json: string) {
```
Comment (swankjesse) [inline misk/web/tabs/config/src/components/ConfigComponent.tsx]: neat! this is the kind of thing that really wants unit tests!

File: misk/web/tabs/config/src/components/ConfigComponent.tsx [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+        case "\,":
+          break
+        default:
+          final += c
+      }
+    }
+    return(final)
+  }
+
+  oldToYaml(json: string) {
```
Comment (swankjesse) [inline misk/web/tabs/config/src/components/ConfigComponent.tsx]: what’s _old_ ? Could ya add a comment?

File: misk/web/tabs/config/src/components/ConfigComponent.tsx [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+    return(json.split(":{").join(":\n \ \ ").split(",").join("\n").split("}").join("\n").split("{").join("").split("\"").join(""))
+  }
+
+  formattedFile(file: IConfigFile) {
+    if (file.name === "live-config.yaml") {
+      console.log("toYaml:", this.toYaml(file.file))
+          return this.toYaml(file.file)
```
Comment (swankjesse) [inline misk/web/tabs/config/src/components/ConfigComponent.tsx]: nit: weird indent

File: misk/web/tabs/config/src/sagas/configSaga.ts [hunk 7]
```diff
@@ -14,6 +14,8 @@ import {
   CONFIG, dispatchConfig, IAction, IActionType
 } from "../actions"
 
+const dateFormat = "YYYY-MM-DD HH:mm:ss"
```
Comment (swankjesse) [inline misk/web/tabs/config/src/sagas/configSaga.ts]: YASSS my fave

Comment (swankjesse) [reply misk/web/tabs/config/src/sagas/configSaga.ts]: Presumably this is local time? Doesn’t matter for this use case, but sometimes we’re mixing UTC and local and it’s clumsy. I have no solutions for that

File: misk/web/tabs/loader/src/containers/LoaderContainer.tsx [hunk 8]
```diff
@@ -1,58 +1,59 @@
...
 import { Route, Switch } from "react-router"
 import { Link } from "react-router-dom"
 import { dispatchLoader } from "../actions"
 import { MountingDivComponent, ScriptComponent } from "../components"
-import { IState } from "../reducers"
+import { ILoaderState, IState } from "../reducers"
+import { IMultibinder } from "../utils/binder"
+
+interface IMiskTabs {
+  [app:string]: any
```
Comment (swankjesse) [inline misk/web/tabs/loader/src/containers/LoaderContainer.tsx]: I want you to explain me this syntax later . . . it’s the first one I’ve seen that isn’t self-explanatory

File: misk/web/tabs/loader/src/containers/LoaderContainer.tsx [hunk 9]
```diff
@@ -1,58 +1,59 @@
...
 class LoaderContainer extends React.Component<ILoaderProps> {
-  componentDidMount() {
-    this.props.getComponentsAndTabs()
+  async componentDidMount() {
+    this.props.getTabs()
   }
 
   buildTabRouteMountingDiv(key: any, tab: IMiskAdminTab) {
-    return(<Route key={key} path={`/_admin/test/${tab.slug}`} render={() => <MountingDivComponent key={key} tab={tab}/>}/>)
+    return(<Route key={key} path={`/_admin/${tab.slug}/`} render={() => <MountingDivComponent key={key} tab={tab}/>}/>)
```
Comment (swankjesse) [inline misk/web/tabs/loader/src/containers/LoaderContainer.tsx]: yess!

File: misk/web/tabs/loader/src/sagas/loaderSaga.tsx [hunk 10]
```diff
@@ -13,6 +13,17 @@ import { all, call, put, takeEvery, takeLatest } from "redux-saga/effects"
 import {
   dispatchLoader, IAction, IActionType, LOADER
 } from "../actions"
+import { IMultibinder } from "../utils/binder"
+
+function * handleCacheTabEntries (action: IAction<IActionType, { MiskBinder: IMultibinder}>) {
+  const { MiskBinder } = action.payload
+    yield put(dispatchLoader.success({ adminTabComponents: { blah: "blah"} , staleTabCache: false }))
+    if (MiskBinder) {
+    yield put(dispatchLoader.success({ adminTabComponents: MiskBinder.TabEntry, staleTabCache: false }))
```
Comment (swankjesse) [inline misk/web/tabs/loader/src/sagas/loaderSaga.tsx]: nit: more weird indentation

File: misk/web/tabs/loader/src/sagas/loaderSaga.tsx [hunk 11]
```diff
@@ -13,6 +13,17 @@ import { all, call, put, takeEvery, takeLatest } from "redux-saga/effects"
 import {
   dispatchLoader, IAction, IActionType, LOADER
 } from "../actions"
+import { IMultibinder } from "../utils/binder"
+
+function * handleCacheTabEntries (action: IAction<IActionType, { MiskBinder: IMultibinder}>) {
+  const { MiskBinder } = action.payload
+    yield put(dispatchLoader.success({ adminTabComponents: { blah: "blah"} , staleTabCache: false }))
```
Comment (swankjesse) [inline misk/web/tabs/loader/src/sagas/loaderSaga.tsx]: what’s a tab cache?

File: misk/web/tabs/loader/src/sagas/loaderSaga.tsx [hunk 12]
```diff
@@ -63,11 +74,21 @@ function * handleGetOneComponent (action: IAction<IActionType, { tab: IMiskAdmin
   }
 }
 
+function * handleRegisterComponent (action: IAction<IActionType, { name: string, Component: any }>) {
+  const { name } = action.payload
+  const { Component } = action.payload
+  console.log("saga")
```
Comment (swankjesse) [inline misk/web/tabs/loader/src/sagas/loaderSaga.tsx]: ?

File: misk/web/tabs/loader/src/containers/LoaderContainer.tsx [hunk 13]
```diff
@@ -1,58 +1,59 @@
...
   render() {
-    const { adminTabs } = this.props.loader
-    const { adminTabComponents } = this.props.loader
+    const { adminTabs, adminTabComponents, staleTabCache } = this.props.loader
     if (adminTabs) {
       const tabRouteDivs = Object.entries(adminTabs).map(([key,tab]) => this.buildTabRouteMountingDiv(key, tab))
-      const tabLinks = Object.entries(adminTabs).map(([key,tab]) => <Link key={key} to={`/_admin/${tab.slug}`}>{tab.name}<br/></Link>)
-      console.log(adminTabs, adminTabComponents)
+      const tabLinks = Object.entries(adminTabs).map(([key,tab]) => <Link key={key} to={`/_admin/${tab.slug}/`}>{tab.name}<br/></Link>)
+      console.log(staleTabCache, (window as any).MiskBinders, adminTabs, adminTabComponents)
```
Comment (terrilldent) [inline misk/web/tabs/loader/src/containers/LoaderContainer.tsx]: intentional console.log?

File: misk/web/tabs/config/src/index.tsx [hunk 14]
```diff
@@ -10,13 +10,23 @@ import App from "./App"
...
+export interface IWindow extends Window {
+  __REDUX_DEVTOOLS_EXTENSION_COMPOSE__: any
+  Misk: {
+    Binder: any
+    History: any
+  }
+}
+const Window = window as IWindow
+
+Window.Misk.History = Window.Misk.History || createBrowserHistory()
```
Comment (terrilldent) [inline misk/web/tabs/config/src/index.tsx]: I wish we could pass / provide this to the sub components instead.

File: misk/web/tabs/config/src/components/ConfigComponent.tsx (hunk 3 above)
Comment (adrw) [reply misk/web/tabs/config/src/components/ConfigComponent.tsx]: This entire part of the ConfigComponent is still very rough/in progress. Trying to figure out how to parse the Jackson JsonNode.toString back to YAML using NPM packages or basic parsing.

File: misk/web/tabs/config/src/components/ConfigComponent.tsx (hunk 4 above)
Comment (adrw) [reply misk/web/tabs/config/src/components/ConfigComponent.tsx]: Absolutely! #372

File: misk/web/tabs/config/src/components/ConfigComponent.tsx (hunk 5 above)
Comment (adrw) [reply misk/web/tabs/config/src/components/ConfigComponent.tsx]: old is the first iteration I did a few weeks ago that was very rough join/split chaining that worked well enough for demo but isn't a long term solution.

File: misk/web/tabs/config/src/sagas/configSaga.ts (hunk 7 above)
Comment (adrw) [reply misk/web/tabs/config/src/sagas/configSaga.ts]: I put this in the demo just to visually show that the Config shown on screen was updated as of a certain point. We should definitely put this in UTC to match existing paradigm. I think DayJS can do this. #427

File: misk/web/tabs/loader/src/containers/LoaderContainer.tsx (hunk 8 above)
Comment (adrw) [reply misk/web/tabs/loader/src/containers/LoaderContainer.tsx]: For sure! Typescript interfaces took a bit of getting used to/

File: misk/web/tabs/loader/src/sagas/loaderSaga.tsx (hunk 11 above)
Comment (adrw) [reply misk/web/tabs/loader/src/sagas/loaderSaga.tsx]: Currently keeping a cache of each tab's code in the React/Redux state. Was part of testing a hypothesis that we could render a component given their compiled code.

File: misk/web/tabs/config/src/components/ConfigComponent.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (adrw) [reply misk/web/tabs/config/src/components/ConfigComponent.tsx]: Absolutely! This entire part of the ConfigComponent is still very rough/in progress. Trying to figure out how to parse the Jackson JsonNode.toString back to YAML using NPM packages or basic parsing.

File: misk/web/tabs/config/src/components/ConfigComponent.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (adrw) [reply misk/web/tabs/config/src/components/ConfigComponent.tsx]: old is the first iteration I did a few weeks ago that was very rough join/split chaining that worked well enough for demo but isn't a long term solution.

File: misk/web/tabs/config/src/index.tsx (hunk 14 above)
Comment (adrw) [reply misk/web/tabs/config/src/index.tsx]: Would definitely be preferable once we find a way to render them without compile time access to the tab components.

File: misk/src/main/kotlin/misk/web/proxy/WebProxyEntry.kt (hunk 1 above)
Comment (adrw) [reply misk/src/main/kotlin/misk/web/proxy/WebProxyEntry.kt]: Will submit another PR addressing this #419

File: misk/web/tabs/config/src/components/ConfigComponent.tsx (hunk 2 above)
Comment (adrw) [reply misk/web/tabs/config/src/components/ConfigComponent.tsx]: 💯 https://www.npmjs.com/package/styled-components

**Pull request merged**